src/unauthenticated_stream/mod.rs

```diff
@@ -0,0 +1,224 @@
+    (key.key.algorithm.xor_keystream)(&key.key.ctx_buf, nonce, in_out)?;
+    Ok(in_out.len())
+}
+
+/// `DecryptingKey` and `EncryptingKey` are type-safety wrappers around `Key`,
+/// which does all the actual work via the C stream cipher interface.
+///
+/// C analog: `EVP_CIPHER_CTX`
+struct Key {
+    ctx_buf: [u64; KEY_CTX_BUF_ELEMS],
+    algorithm: &'static Algorithm,
+}
+
+const KEY_CTX_BUF_ELEMS: usize = (KEY_CTX_BUF_LEN + 7) / 8;
+
+const KEY_CTX_BUF_LEN: usize = chacha::KEY_LEN_IN_BYTES;
+
+impl Key {
+    fn new(algorithm: &'static Algorithm, key_bytes: &[u8]) -> Result<Self, error::Unspecified> {
+        if key_bytes.len() != algorithm.key_len() {
+            return Err(error::Unspecified);
+        }
+
+        let mut r = Key {
+            algorithm,
+            ctx_buf: [0; KEY_CTX_BUF_ELEMS],
+        };
+
+        init::init_once();
+        {
+            let ctx_buf_bytes = polyfill::slice::u64_as_u8_mut(&mut r.ctx_buf);
+            (r.algorithm.init)(ctx_buf_bytes, key_bytes)?;
+        }
+
+        Ok(r)
+    }
+
+    /// The key's unauthenticated stream cipher algorithm.
+    #[inline(always)]
+    fn algorithm(&self) -> &'static Algorithm { self.algorithm }
+}
+
+/// An unauthenticated stream cipher Algorithm.
+///
+/// C analog: `EVP_CIPHER`
+///
+/// Go analog:
+///     [`crypto.cipher.Stream`](https://golang.org/pkg/crypto/cipher/#Stream)
+pub struct Algorithm {
+    init: fn(ctx_buf: &mut [u8], key: &[u8]) -> Result<(), error::Unspecified>,
+
+    xor_keystream: fn(ctx: &[u64; KEY_CTX_BUF_ELEMS], nonce: &[u8; NONCE_LEN],
+                      in_out: &mut [u8]) -> Result<(), error::Unspecified>,
+
+    key_len: usize,
+    id: AlgorithmID,
+
+    /// Use `max_input_len!()` to initialize this.
+    // TODO: Make this `usize`.
+    max_input_len: u64,
+}
+
+/// TODO: Make this a `const fn` when those become stable.
+macro_rules! max_input_len {
+    ($block_len:expr, $overhead_blocks_per_nonce:expr) => {
+        // Each of our AEADs use a 32-bit block counter so the maximum is the
+        // largest input that will not overflow the counter.
+        (((1u64 << 32) - $overhead_blocks_per_nonce) * $block_len)
+    }
+}
+
+impl Algorithm {
+    /// The length of the key.
+    ///
+    /// C analog: `EVP_CIPHER_key_length`
+    #[inline(always)]
+    pub fn key_len(&self) -> usize { self.key_len }
+
+    /// The length of the nonces.
+    ///
+    /// C analog: `EVP_CIPHER_nonce_length`
+    ///
+    /// Go analog:
+    /// [`crypto.cipher.Stream.NonceSize`](https://golang.org/pkg/crypto/cipher/#Stream)
+    #[inline(always)]
+    pub fn nonce_len(&self) -> usize { NONCE_LEN }
+}
+
+derive_debug_from_field!(Algorithm, id);
+
+#[derive(Debug, Eq, PartialEq)]
+enum AlgorithmID {
+    CHACHA20,
+}
+
+impl PartialEq for Algorithm {
+    fn eq(&self, other: &Self) -> bool { self.id == other.id }
+}
+
+impl Eq for Algorithm {}
+
+// All the stream ciphers we support use 128-bit nonces.
+const NONCE_LEN: usize = 128 / 8;
```

Comment:
Let's avoid duplicating this definition too.

Reply:
This is different from `aead` (16 bytes here vs. 12 bytes in aead).